src/puppetlabs/puppetdb/cli/benchmark.clj

```diff
@@ -472,27 +429,148 @@
+        shutdown-thread (promise)  ;; Because no letrec...
+        stopping? (AtomicBoolean.)
+        stop (fn []
+               (if (.getAndSet stopping? true)
+                 true
+                 (do
+                   ;; Close narrow race between hook install and
+                   ;; fill-queue start.  Give up after 3s and shutdown
+                   ;; anyway (perhaps more noisily).
+                   (doseq [i (range 30)
+                           :while (= Thread$State/NEW (.getState fill-queue))]
+                     (Thread/sleep 100))
+                   (try
+                     ;; This removal will only succeed when stop is
+                     ;; called directly (not as a hook), which is
+                     ;; currently only during testing.  A normal
+                     ;; benchmark process solely relies on
+                     ;; C-c to shut down, in which case we'll see the
+                     ;; IllegalStateException here.
+                     (.removeShutdownHook (Runtime/getRuntime) @shutdown-thread)
+                     (catch IllegalStateException _
+                       nil))
+                   ;; This is a mess because we want to make sure we
+                   ;; always try to close all the channels, and *always*
+                   ;; call delete-dir.
+                   (let [close-error?
+                         (try
+                           (not-every? identity
+                                       (map #(try
+                                               (close! %)
+                                               ;; drain channel
+                                               (while (<!! %) true)
+                                               true
+                                               (catch Exception ex
+                                                 (println-err ex)
+                                                 ex
+                                                 false))
+                                            channels))
+                           (catch Exception ex
+                             (println-err ex)
+                             ex
+                             false))
+                         join-error?
+                         (try
+                           (.join fill-queue 500)
+                           (when (.isAlive fill-queue)
+                             (println-err (trs "Waiting up to 10s for shutdown"))
+                             ;; Do we need to make this longer b/c CI mess?
+                             (.join fill-queue 10000))
+                           false
+                           (catch Exception ex
+                             (println-err ex)
+                             true))]
+                     (fs/delete-dir tmpdir)
+                     (and (not close-error?)
+                          (not join-error?)
+                          (not (.isAlive fill-queue)))))))]
```

Comment:
This all seems kind of complicated to place in a let block. I wonder if it would be cleaner to move it outside the let?